akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -0,0 +1,218 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery
+
+import java.nio.charset.Charset
+import java.nio.{ ByteBuffer, ByteOrder }
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.{ ActorRef, Address, AddressFromURIString }
+import akka.io.DirectByteBufferPool
+import akka.remote.RemoteActorRef
+import org.agrona.concurrent.ManyToManyConcurrentArrayQueue
+
+/**
+ * INTERNAL API
+ */
+private[remote] class OutOfBuffersException extends RuntimeException("Out of usable ByteBuffers")
+
+/**
+ * INTERNAL API
+ */
+private[remote] class PayloadBufferPool(maximumPayload: Int, maximumBuffers: Int) {
+
+  private val allocateBudget = new AtomicInteger(maximumBuffers)
+
+  private val directByteBufferPool = new DirectByteBufferPool(maximumPayload, maximumBuffers)
+  private val availableBuffers = new ManyToManyConcurrentArrayQueue[EnvelopeBuffer](maximumBuffers)
+
+  def acquire(): EnvelopeBuffer = {
+    val buf = availableBuffers.poll()
+
+    if (buf ne null) {
+      buf
+    } else if (allocateBudget.decrementAndGet() >= 0) {
+      val newBuf = new EnvelopeBuffer(directByteBufferPool.acquire())
+      newBuf.byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+      newBuf
+    } else
+      throw new OutOfBuffersException
+  }
+
+  def release(buffer: EnvelopeBuffer) = {
+    availableBuffers.add(buffer)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] object EnvelopeBuffer {
+
+  val TagTypeMask = 0xFF000000
+  val TagValueMask = 0x0000FFFF
+
+  val VersionOffset = 0
+  val UidOffset = 4
+  val SenderActorRefTagOffset = 8
+  val RecipientActorRefTagOffset = 12
+  val LiteralsSectionOffset = 16
+
+  val UsAscii = Charset.forName("US-ASCII")
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] trait LiteralCompressionTable {
+
+  def compressActorRef(ref: String): Int
+  def decompressActorRef(idx: Int): String
+
+}
+
+object HeaderBuilder {
+  def apply(compressionTable: LiteralCompressionTable): HeaderBuilder = new HeaderBuilderImpl(compressionTable)
+}
+
+/**
+ * INTERNAL API
+ */
+sealed trait HeaderBuilder {
+  def version_=(v: Int): Unit
+  def version: Int
+
+  def uid_=(u: Int): Unit
+  def uid: Int
+
+  def senderActorRef_=(ref: String): Unit
+  def senderActorRef: String
+
+  def recipientActorRef_=(ref: String): Unit
+  def recipientActorRef: String
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class HeaderBuilderImpl(val compressionTable: LiteralCompressionTable) extends HeaderBuilder {
```

Comment:
"do we need an interface" reminds me of an excellent issue on Gueva I read the other day (not directly related, but somewhat):
https://github.com/google/guava/issues/1268

Very interesting that by having too many arities optimised as raw impls for Lists they blew performance because it seems bimorphic actually is a thing, and better than megamorphic (I had always assumed it get's bad at 2 impls already). Only found it last week, sorry if you've seen it before :)